File: AspNetMvcUnleashed/0672329980_source/Chapter21/CS/UnleashedBlog/UnleashedBlog.Tests/Controllers/ArchiveControllerTests.cs
using System;
using System.Collections.Generic;
using System.Web.Mvc;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using UnleashedBlog.Controllers;
using UnleashedBlog.Models;
using UnleashedBlog.Tests.Models;
using UnleashedBlog.Tests.Factories;

namespace UnleashedBlog.Tests.Controllers
{
    [TestClass]
    public class ArchiveControllerTests
    {
        private ControllerFactory _controllerFactory;
        private ArchiveController _archiveController;


        /// <summary>
        /// Because we use the Archive controller in every test, let's just create
        /// one automatically.
        /// </summary>
        [TestInitialize]
        public void Initialize()
        {
            _controllerFactory = new ControllerFactory();
            _archiveController = _controllerFactory.GetArchiveController();
        }


        /// <summary>
        /// Verifies that Index() action returns blog entries that match a specific year.
        /// </summary>
        [TestMethod]
        public void IndexReturnsBlogEntriesByYear()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 11, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2011, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(null, 2010, null, null);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }

        /// <summary>
        /// Verifies that Index() action returns blog entries that match a specific month.
        /// </summary>
        [TestMethod]
        public void IndexReturnsBlogEntriesByMonth()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 11, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(null, 2010, 12, null);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }

        /// <summary>
        /// Verifies that Index() action returns blog entries that match a specific day.
        /// </summary>
        [TestMethod]
        public void IndexReturnsBlogEntriesByDay()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(null, 2010, 12, 25);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }

        /// <summary>
        /// Verifies that Index() action returns blog entries that match a specific name.
        /// </summary>
        [TestMethod]
        public void IndexReturnsBlogEntryByName()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();
            bController.Create(BlogEntryFactory.GetWithName("Test 1"));
            bController.Create(BlogEntryFactory.GetWithName("Test 2"));

            // Act
            var result = (ViewResult)_archiveController.Details(2010, 12, 25, "Test-2");

            // Assert
            var blogEntry = (BlogEntry)result.ViewData.Model;
            Assert.AreEqual("Test-2", blogEntry.Name);
        }

        //____________________
        // Ajax Tests
        //____________________

        /// <summary>
        /// Verifies that Index() action returns a PartialViewResult.
        /// </summary>
        [TestMethod]
        public void Index_AjaxReturnsPartialViewResult()
        {
            // Act
            var result = _archiveController.Index_Ajax(0, 2010, 12, 25);

            // Assert
            Assert.IsInstanceOfType(result, typeof(PartialViewResult));
        }


    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter21/VB/UnleashedBlog.Tests/Controllers/ArchiveControllerTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports System.Web.Mvc

<TestClass()> _
 Public Class ArchiveControllerTests
    Private _controllerFactory As ControllerFactory
    Private _archiveController As ArchiveController


    ''' <summary>
    ''' Because we use the Archive controller in every test, let's just create
    ''' one automatically.
    ''' </summary>
    <TestInitialize()> _
    Public Sub Initialize()
        _controllerFactory = New ControllerFactory()
        _archiveController = _controllerFactory.GetArchiveController()
    End Sub


    ''' <summary>
    ''' Verifies that Index() action returns blog entries that match a specific year.
    ''' </summary>
    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByYear()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 11, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2011, 12, 26)))

        ' Act
        Dim result = CType(_archiveController.Index(Nothing, 2010, Nothing, Nothing), ViewResult)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub

    ''' <summary>
    ''' Verifies that Index() action returns blog entries that match a specific month.
    ''' </summary>
    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByMonth()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 11, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 26)))

        ' Act
        Dim result = CType(_archiveController.Index(Nothing, 2010, 12, Nothing), ViewResult)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub

    ''' <summary>
    ''' Verifies that Index() action returns blog entries that match a specific day.
    ''' </summary>
    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByDay()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 26)))

        ' Act
        Dim result = CType(_archiveController.Index(Nothing, 2010, 12, 25), ViewResult)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub

    ''' <summary>
    ''' Verifies that Index() action returns blog entries that match a specific name.
    ''' </summary>
    <TestMethod()> _
    Public Sub IndexReturnsBlogEntryByName()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()
        bController.Create(BlogEntryFactory.GetWithName("Test-1"))
        bController.Create(BlogEntryFactory.GetWithName("Test-2"))

        ' Act
        Dim result = CType(_archiveController.Details(2010, 12, 25, "Test-2"), ViewResult)

        ' Assert
        Dim blogEntry = CType(result.ViewData.Model, BlogEntry)
        Assert.AreEqual("Test-2", blogEntry.Name)
    End Sub

    '____________________
    ' Ajax Tests
    '____________________

    ''' <summary>
    ''' Verifies that Index() action returns a PartialViewResult.
    ''' </summary>
    <TestMethod()> _
    Public Sub Index_AjaxReturnsPartialViewResult()
        ' Act
        Dim result = _archiveController.Index_Ajax(0, 2010, 12, 25)

        ' Assert
        Assert.IsInstanceOfType(result, GetType(PartialViewResult))
    End Sub


End Class
